src/optimizer/cost_calculator.cpp

```diff
@@ -50,14 +53,73 @@ void CostCalculator::Visit(UNUSED_ATTRIBUTE const PhysicalIndexScan *op) {
   auto table_stats = std::dynamic_pointer_cast<TableStats>(
       StatsStorage::GetInstance()->GetTableStats(
           op->table_->GetDatabaseOid(), op->table_->GetTableOid(), txn_));
-  if (table_stats->GetColumnCount() == 0 || table_stats->num_rows == 0) {
+  auto index_scan_rows = (double)table_stats->num_rows;
+  if (table_stats->GetColumnCount() == 0 || index_scan_rows == 0) {
     output_cost_ = 0.f;
     return;
   }
+  auto index_object = op->table_->GetIndexObject(op->index_id);
+  const auto &key_attr_list = index_object->GetKeyAttrs();
+  // Loop over index to retrieve helpful index columns
+  // Consider all predicates that could be accelerated by the index,
+  // i.e. till the first column with no equality predicate on it
+  // index cols (a, b, c)
+  // example1 : predicates(a=1 AND b=2 AND c=3) index helps on both a, b and c
+  // example2 : predicates(a<1 AND b<=2 and c<3) index helps on only a
+  // example3 : predicates(a=1 AND b>2 AND c>3) index helps on a and b
+  bool has_non_equality_pred = false;
+  for (size_t idx = 0; idx < key_attr_list.size(); ++idx) {
+    // If index cannot further reduce scan range, break
+    if (idx == op->key_column_id_list.size() ||
+        key_attr_list[idx] != op->key_column_id_list[idx]) {
```

Comment:
@chenboy It looks like this thing requires the `key_column_id_list` has exactly the same order as `key_attr_list`? But in fact, an index with col(a, b, c) can also help a query with predicates(b=2 AND a=1 AND c=3), right?